tools/script_manager.lua

```diff
@@ -0,0 +1,764 @@
+
+-- command separator
+local CS = dt.configuration.running_os == "windows" and "&" or ";"
+
+local LUA_DIR = dt.configuration.config_dir .. PS .. "lua"
+local LUA_SCRIPT_REPO = "https://github.com/darktable-org/lua-scripts.git"
+
+dt.print_log("LUA_DIR is " .. LUA_DIR)
+
+-- - - - - - - - - - - - - - - - - - - - - - - - 
+-- N A M E  S P A C E
+-- - - - - - - - - - - - - - - - - - - - - - - - 
+
+local script_manager = {}
+local sm = script_manager
+
+-- - - - - - - - - - - - - - - - - - - - - - - - 
+-- F U N C T I O N S
+-- - - - - - - - - - - - - - - - - - - - - - - - 
+
+local function prequire(script)
+  dt.print_log("Loading " .. script)
+  local status, lib = pcall(require, script)
+  if status then
+    dt.print_log("Loaded " .. script)
+  else
+    dt.print_error("Error loading " .. script)
+    dt.print_error(lib)
+  end
+  return status, lib
+end
+
+local function update_combobox_choices(combobox, choice_table, selected)
+  local items = #combobox
+  local choices = #choice_table
+  for i, name in ipairs(choice_table) do 
+    combobox[i] = name
+  end
+  if choices < items then
+    for j = items, choices + 1, -1 do
+      combobox[j] = nil
+    end
+  end
+  combobox.value = selected
+end
+
+local function install_scripts()
+  local result = false
+
+  if df.check_if_file_exists(LUA_DIR) then
+    if df.check_if_file_exists(LUA_DIR .. ".orig") then
+      if dt.configuration.running_os == "windows" then
+        os.execute("rmdir /s " .. LUA_DIR .. ".orig")
+      else
+        os.execute("rm -rf " .. LUA_DIR .. ".orig")
+      end
+    end
+    os.rename(LUA_DIR, LUA_DIR .. ".orig")
+  end
+
+  local git = df.check_if_bin_exists("git")
+
+  if not git then
+    dt.print("ERROR: git not found.  Install or specify the location of the git executable.")
+    return
+  end
+
+  local git_command = git .. " clone " .. sm.lua_repository .. " " .. LUA_DIR
+  dt.print_log("install git command is " .. git_command)
+
+  result = os.execute(git_command)
+
+  if not df.check_if_file_exists(LUA_DIR .. PS .. "downloads") then
+    os.execute("mkdir " .. LUA_DIR .. PS .. "downloads")
+  end
+
+  return result
+end
+
+local function update_scripts()
+  local result = false
+
+  local git = df.check_if_bin_exists("git")
+
+  if not git then
+    dt.print("ERROR: git not found.  Install or specify the location of the git executable.")
+    return
+  end
+
+  local git_command = "cd " .. LUA_DIR .. " " .. CS .. " " .. git .. " pull"
+  dt.print_log("update git command is " .. git_command)
+
+  if dt.configuration.running_os == "windows" then
+    result = dtsys.windows_command(git_command)
+  else
+    result = os.execute(git_command)
+  end
+
+  if dt.preferences.read("script_manager", "use_lua_scripts_version", "bool") and 
+     dt.configuration.running_os == "windows" then
+    use_lua_scripts_version()
+  end
+
+  return result
+end
+
+local function add_script_data(script_file)
+
+  local pattern = "(.-)/(.-)(([^\\/]-)%.?([^%.\\/]*))$"
```

Comment:
Please add documentation to pattern like this. This looks like write only code. ;)

Reply:
I think this pattern came from you originally.  :-)

Reply:
That may be, but it is still evil code. ;)